src/NServiceBus.Transport.SqlServer/Receiving/MessagePump.cs

```diff
@@ -125,55 +124,59 @@ async Task ProcessMessages()
         {
-            while (!cancellationTokenSource.IsCancellationRequested)
+            while (!messagePumpCancellationToken.IsCancellationRequested)
             {
-                var messageCount = await queuePeeker.Peek(inputQueue, peekCircuitBreaker, cancellationToken).ConfigureAwait(false);
+                var messageCount = await queuePeeker.Peek(inputQueue, peekCircuitBreaker, messagePumpCancellationToken).ConfigureAwait(false);
 
                 if (messageCount == 0)
                 {
                     continue;
                 }
 
-                if (cancellationTokenSource.IsCancellationRequested)
+                if (messagePumpCancellationToken.IsCancellationRequested)
                 {
                     return;
                 }
 
                 // We cannot dispose this token source because of potential race conditions of concurrent receives
-                var loopCancellationTokenSource = new CancellationTokenSource();
+                var stopBatchCancellationSource = new CancellationTokenSource();
 
                 // If the receive or peek circuit breaker is triggered start only one message processing task at a time.
                 var maximumConcurrentReceives = receiveCircuitBreaker.Triggered || peekCircuitBreaker.Triggered ? 1 : messageCount;
 
                 for (var i = 0; i < maximumConcurrentReceives; i++)
                 {
-                    if (loopCancellationTokenSource.Token.IsCancellationRequested)
+                    if (messagePumpCancellationToken.IsCancellationRequested || stopBatchCancellationSource.IsCancellationRequested)
                     {
                         break;
                     }
 
-                    await concurrencyLimiter.WaitAsync(cancellationToken).ConfigureAwait(false);
+                    await concurrencyLimiter.WaitAsync(messagePumpCancellationToken).ConfigureAwait(false);
 
-                    _ = InnerReceive(loopCancellationTokenSource);
+                    _ = InnerReceive(stopBatchCancellationSource, messageProcessingCancellationTokenSource.Token);
                 }
             }
         }
 
-        async Task InnerReceive(CancellationTokenSource loopCancellationTokenSource)
+        async Task InnerReceive(CancellationTokenSource stopBatch, CancellationToken messageProcessingCancellationToken)
         {
             try
             {
                 // We need to force the method to continue asynchronously because SqlConnection
                 // in combination with TransactionScope will apply connection pooling and enlistment synchronous in ctor.
                 await Task.Yield();
 
-                await receiveStrategy.ReceiveMessage(loopCancellationTokenSource)
+                await receiveStrategy.ReceiveMessage(stopBatch, messageProcessingCancellationToken)
                     .ConfigureAwait(false);
 
                 receiveCircuitBreaker.Success();
             }
+            catch (OperationCanceledException)
```

Comment:
Should this be similar to the [learning transport](https://github.com/Particular/NServiceBus/blob/ed4c0cd916572864a424d29b6835b99a0f2772c4/src/NServiceBus.Core/Transports/Learning/LearningTransportMessagePump.cs#L311)?

```suggestion
            catch (OperationCanceledException) when (stopBatch.IsCancellationRequested || messageProcessingCancellationToken.IsCancellationRequested)
```